src/vos/vos_obj_cache.c

```diff
@@ -194,124 +196,124 @@ vos_obj_release(struct daos_lru_cache *occ, struct vos_object *obj)
-
-			rc = vos_dtx_check_availability(umm, vos_cont2hdl(cont),
-						obj->obj_df->vo_dtx,
-						umem_ptr2off(umm, obj->obj_df),
-						intent, DTX_RT_OBJ);
-			if (rc < 0)
-				D_GOTO(failed, rc);
+	rc = daos_lru_ref_hold(occ, &lkey, sizeof(lkey), cont, &lret);
+	if (rc)
+		D_GOTO(failed_2, rc);
 
-			if (rc != ALB_UNAVAILABLE) {
-				if (obj->obj_incarnation == 0)
-					obj->obj_epoch = epoch;
+	obj = container_of(lret, struct vos_object, obj_llink);
 
-				goto out;
-			}
-		}
+	if (obj->obj_zombie)
+		D_GOTO(failed, rc = -DER_AGAIN);
 
-		D_DEBUG(DB_IO, "Evict obj ["DF_U64" -> "DF_U64"]\n",
-			obj->obj_epoch, epoch);
+	if (intent == DAOS_INTENT_KILL) {
+		if (vos_obj_refcount(obj) > 2)
+			D_GOTO(failed, rc = -DER_BUSY);
 
-		/* NB: we don't expect user wants to access many versions
-		 * of the same object at the same time, so just evict the
-		 * unmatched version from the cache, then populate the cache
-		 * with the demanded versoin.
-		 */
+		/* no one else can hold it */
+		obj->obj_zombie = true;
 		vos_obj_evict(obj);
-		vos_obj_release(occ, obj);
+		if (obj->obj_df)
+			goto out; /* Ok to delete */
 	}
 
-	D_DEBUG(DB_TRACE, "%s Got empty obj "DF_UOID" in epoch="DF_U64"\n",
-		no_create ? "find" : "find/create", DP_UOID(oid), epoch);
+	if (obj->obj_df)
+		goto check_object;
 
+	 /* newly cached object */
+	D_DEBUG(DB_TRACE, "%s Got empty obj "DF_UOID" epr="DF_U64"-"DF_U64"\n",
+		no_create ? "find" : "find/create", DP_UOID(oid), epr->epr_lo,
+		epr->epr_hi);
+
+	obj->obj_sync_epoch = 0;
 	if (no_create) {
-		rc = vos_oi_find(cont, oid, epoch, intent, &obj->obj_df);
+		rc = vos_oi_find(cont, oid, &obj->obj_df);
 		if (rc == -DER_NONEXIST) {
 			D_DEBUG(DB_TRACE, "non exist oid "DF_UOID"\n",
 				DP_UOID(oid));
-			obj->obj_sync_epoch = 0;
-			rc = 0;
-		} else if (rc == 0) {
-			obj->obj_sync_epoch = obj->obj_df->vo_sync;
+			goto failed;
 		}
 	} else {
-		rc = vos_oi_find_alloc(cont, oid, epoch, intent, &obj->obj_df);
+		rc = vos_oi_find_alloc(cont, oid, epr->epr_hi, false,
+				       &obj->obj_df);
 		D_ASSERT(rc || obj->obj_df);
 	}
 
-	if (rc)
+	if (rc != 0)
 		goto failed;
 
 	if (!obj->obj_df) {
 		D_DEBUG(DB_TRACE, "nonexistent obj "DF_UOID"\n",
 			DP_UOID(oid));
-		if (intent == DAOS_INTENT_KILL) {
-			D_GOTO(failed, rc = -DER_NONEXIST);
-		}
+		D_GOTO(failed, rc = -DER_NONEXIST);
 		goto out;
 	}
+	obj->obj_sync_epoch = obj->obj_df->vo_sync;
+check_object:
+	if (intent == DAOS_INTENT_KILL || intent == DAOS_INTENT_PUNCH)
+		goto out;
 
-	D_ASSERTF((obj->obj_df->vo_oi_attr & VOS_OI_PUNCHED) == 0 ||
-		  epoch <= obj->obj_df->vo_latest,
-		  "e="DF_U64", p="DF_U64"\n", epoch,
-		  obj->obj_df->vo_latest);
+	if (no_create)
+		goto log_fetch;
 
-	obj->obj_incarnation = obj->obj_df->vo_incarnation;
+	vos_ilog_desc_cbs_init(&cbs, vos_cont2hdl(cont));
+	rc = ilog_open(vos_cont2umm(cont), &obj->obj_df->vo_ilog, &cbs, &loh);
+	if (rc != 0)
+		goto failed;
+
+	rc = ilog_update(loh, NULL, epr->epr_hi, false);
```

Comment:
Ilog_update is called because we don’t know if the object exists at the epoch we are updating.  If it does and it’s committed, this will be a noop internally.